File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.java
class Solution {
    public String frequencySort(String s) {
        Map<Character, Integer> charCount = new HashMap<>();
        for (char c : s.toCharArray()) {
            charCount.put(c, charCount.getOrDefault(c, 0) + 1);
        }
        
        PriorityQueue<Map.Entry<Character, Integer>> pq = new PriorityQueue<>(
            (a, b) -> b.getValue() - a.getValue() 
        );
        pq.addAll(charCount.entrySet());
        
        StringBuilder result = new StringBuilder();
        while (!pq.isEmpty()) {
            Map.Entry<Character, Integer> entry = pq.poll();
            char c = entry.getKey();
            int count = entry.getValue();
            for (int i = 0; i < count; i++) {
                result.append(c);
            }
        }
        
        return result.toString();
    }
}

// class Solution {
//     public String frequencySort(String s) {
//         Map<Character, Integer> mp = new HashMap<>();
//         for (char c : s.toCharArray()) {
//             mp.put(c, mp.getOrDefault(c, 0) + 1);
//         } 
        
//         List<Character> charList = new ArrayList<>();
//         for (char c : s.toCharArray()) {
//             charList.add(c);
//         }
        
//         Comparator<Character> comparator = (a, b) -> {
//             if (mp.get(a).equals(mp.get(b))) {
//                 return a - b;
//             }
//             return mp.get(b) - mp.get(a);
//         };
        
//         Character[] charArray = charList.toArray(new Character[charList.size()]);
//         Arrays.sort(charArray, comparator);
        
//         StringBuilder sb = new StringBuilder();
//         for (char c : charArray) {
//             sb.append(c);
//         }
        
//         return sb.toString();
//     }
// }

File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.cpp
class Solution {
public:
    typedef pair<char, int> P;

    struct lambda  {
        bool operator()(P &p1, P &p2) {
            return p1.second < p2.second;
        }
    };

    string frequencySort(string s) {
        priority_queue<P, vector<P>, lambda> pq;
        unordered_map<char, int> mp;
        for (char c : s) {
            mp[c]++;
        }
        for (auto &it : mp) {
            pq.push({it.first, it.second});
        }
        string result = "";
        while (!pq.empty()) {
            P temp = pq.top();
            pq.pop();
            result += string(temp.second, temp.first);
        }
        return result;
    }
};

// class Solution {
// public:
//     string frequencySort(string s) {
//         unordered_map<char, int> mp;
//         for (char c : s) {
//             mp[c]++;
//         }
//         auto comparator = [&] (char a, char b) {
//             if (mp[a] == mp[b]) {
//                 return a < b;
//             }
//             return mp[a] > mp[b];
//         };
//         sort(s.begin(), s.end(), comparator);
//         return s;
//     }
// };